### DISCUSSION TITLE
Deprecation advice for navigationDestination(store:destination) on macOS

### DISCUSSION THREAD

**Discussion by johankool** [pointfreeco/swift-composable-architecture#2809] [Q&A]
### Description

This is about a SwiftUI bug that happens on macOS. I report it here because the deprecation advices to use this code.

In [NavigationDestination.swift](https://github.com/pointfreeco/swift-composable-architecture/blob/de63b1d436a4d6fcc8c39f23fe4eb9a10b867a52/Sources/ComposableArchitecture/SwiftUI/NavigationDestination.swift#L24) it says:
```swift
  @available(
    macOS, deprecated: 9999,
    message:
      "Pass a binding of a store to 'navigationDestination(item:)' instead. For more information, see the following article: https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/migratingto1.7#Replacing-navigation-view-modifiers-with-SwiftUI-modifiers]"
  )
```

Unfortunately this functionality is broken on macOS. If the recommended approach is used nothing happens on macOS when navigating.

```swift
// Does nothing on macOS 14
.navigationDestination(item: $store.scope(state: \.destination?.connect, action: \.destination.connect)) { store in
    ConnectView(store: store)
}
```

Staying with the deprecated approach, kinda works, but not really as it breaks perception/observation.

```swift
// Works, but purple warnings
.navigationDestination(store: store.scope(state: \.$destination.connect, action: \.destination.connect)) { store in
    ConnectView(store: store)
}
```

The workaround I found is that I need a bit of glue code in the store to calculate an isPresented binding and use the `navigationDestination(isPresented:destination:)` variant instead.

```swift
// Works with some glue code in the store to derive isConnecting from the current destination
.navigationDestination(isPresented: $store.isConnecting) {
    if let store = store.scope(state: \.destination?.connect, action: \.destination.connect) {
        ConnectView(store: store)
    }
}
```

### Checklist

- [X] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### The Composable Architecture version information

1.8.0

### Destination operating system

macOS 14.3.1

### Xcode version information

Version 15.2 (15C500b)

### Swift Compiler version information

```shell
swift-driver version: 1.87.3 Apple Swift version 5.9.2 (swiftlang-5.9.2.2.56 clang-1500.1.0.2.5)
Target: arm64-apple-macosx14.0
```

**Comment by stephencelis** [comment]
@johankool Another option is to use the polypill we define here:

https://pointfreeco.github.io/swift-composable-architecture/1.8.0/documentation/composablearchitecture/treebasednavigation#Backwards-compatible-availability

Just remove the deprecation availability.

Since this is an Apple bug I'm going to convert to a discussion. We can chat about modifying our deprecation message there, but my worry is I don't think we can validate every Apple API we point folks towards across every platform for every release.
